deep_qa/data/dataset.py

```diff
@@ -0,0 +1,157 @@
+import logging
+from collections import defaultdict
+from typing import Dict, List
+
+import numpy
+import tqdm
+
+from ...common.util import add_noise_to_dict_values
+from . import Instance, Vocabulary
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class Dataset:
+    """
+    A collection of :class:`Instance` objects.  The ``Instances`` have ``Fields``, and the fields
+    could be in an indexed or unindexed state - the ``Dataset`` has methods around indexing the
+    data and converting the data into arrays.
+    """
+    def __init__(self, instances: List[Instance]):
+        """
+        A Dataset just takes a list of instances in its constructor.  It's important that all
+        subclasses have an identical constructor to this (though possibly with different Instance
+        types).  If you change the constructor, you also have to override all methods in this base
+        class that call the constructor, such as `merge()` and `truncate()`.
+        """
+        self.instances = instances
+
+    def truncate(self, max_instances: int):
```

Comment:
Yeah, that's a fair point.  It's my bias towards making things immutable, but `Dataset` is already very much not immutable, so I might as well make this one consistent.  I'll change it.